plugin/trino-kafka-event-listener/src/main/java/io/trino/plugin/eventlistener/kafka/KafkaEventPublisher.java

```diff
@@ -0,0 +1,149 @@
+
+    private final KafkaProducer<String, GenericRecord> kafkaProducer;
+    private final KafkaRecordBuilder kafkaRecordBuilder;
+    private final KafkaEventListenerJmxStats stats;
+
+    KafkaEventPublisher(KafkaEventListenerConfig config, KafkaProducerFactory producerFactory, KafkaEventListenerJmxStats stats)
+            throws Exception
+    {
+        this.stats = stats;
+
+        String startedTopic = config.getCreatedTopicName().orElse("");
+        String completedTopic = config.getCompletedTopicName().orElse("");
+
+        Map<String, String> configOverrides = config.getKafkaClientOverrides();
+        LOG.info("Kafka Event Listener: Creating Kafka publisher (SSL=%s) for topics: %s/%s with excluded fields: %s and kafka config overrides: %s",
+                producerFactory instanceof SSLKafkaProducerFactory, startedTopic, completedTopic, config.getExcludedFields(), configOverrides);
+        kafkaProducer = producerFactory.producer(configOverrides);
+        checkConnectivityToBrokers(config.getPublishCreatedEvent() ? startedTopic : completedTopic, config.getRequestTimeout());
+        kafkaRecordBuilder = new KafkaRecordBuilder(startedTopic, completedTopic, config.getExcludedFields(), metadataProvider(config));
+        LOG.info("Kafka Event Listener: Successfully created Kafka publisher.");
+    }
+
+    private void checkConnectivityToBrokers(String topic, int requestTimeout)
+            throws Exception
+    {
+        try {
+            LOG.debug("Kafka Event Listener: checking connectivity to brokers (fetching partitions for topic=%s).", topic);
+            CompletableFuture<Void> future = CompletableFuture.runAsync(() -> kafkaProducer.partitionsFor(topic));
+            future.get(requestTimeout, TimeUnit.MILLISECONDS);
+            LOG.debug("Kafka Event Listener: connectivity check succeeded.");
+        }
+        catch (Exception e) {
+            LOG.error("Kafka Event Listener: connectivity check failed: %s", e.getMessage());
+            throw e;
+        }
+    }
+
+    private MetadataProvider metadataProvider(KafkaEventListenerConfig config)
+    {
+        if (config.getEnvironmentVariablePrefix().isPresent()) {
+            return new EnvMetadataProvider(config.getEnvironmentVariablePrefix().get());
+        }
+        else {
+            return new NoOpMetadataProvider();
+        }
+    }
+
+    public void publishCompletedEvent(QueryCompletedEvent queryCompletedEvent)
+    {
+        stats.completedEventReceived();
+        String queryId = queryCompletedEvent.getMetadata().getQueryId();
+        LOG.debug("Kafka Event Listener: preparing to send QueryCompletedEvent for query id: %s", queryId);
+        ProducerRecord<String, GenericRecord> record = null;
+        try {
+            record = kafkaRecordBuilder.buildCompletedRecord(queryCompletedEvent);
+        }
+        catch (Exception e) {
+            stats.completedEventBuildFailure();
+            if (LOG.isDebugEnabled()) {
```

Comment:
I would just leave `WARM` since failure to build object seems like something to checkout.

Same for `publishCreatedEvent`